exporter/gpu_stats.py

```diff
@@ -0,0 +1,128 @@
+from base import OSBase
+from collections import defaultdict
+from osclient import get_keystone_session, session_adapter
+from os import environ
+from prometheus_client import CollectorRegistry, generate_latest, Gauge
+import logging
+import re
+
+logger = logging.getLogger(__name__)
+
+GNOCCHI_ENDPOINT = environ['GNOCCHI_ENDPOINT']
+GRANULARITY = 60
+AGGREGATION_METHOD = 'mean'
+LABELS = ['region', 'stat_name', 'gpu_type', 'gpu_index']
+
+
+class GPUStats(OSBase):
+    """Class to report the statistics on NVIDIA GPUs"""
+
+    def __init__(self, oscache, osclient):
+        super(GPUStats, self).__init__(oscache, osclient)
+
+        self.sess = get_keystone_session()
+
+    def build_cache_data(self):
+        """Return list of stats to cache"""
+        cache_stats = []
+
+        for gpu_type, metrics in self.get_metrics_by_gpu_type().items():
+            iter = 0
+
+            for metric, gpu_indices in metrics.items():
+                metric_name = str(metric).split('.')[-1]
+
+                for gpu, metric_ids in gpu_indices.items():
+                    stat = dict(
+                        stat_name=metric_name,
+                        gpu_type=gpu_type,
+                        gpu_index=gpu)
+
+                    stat['stat_value'] = self.get_aggregate(metric_ids)
+                    cache_stats.append(stat)
+
+                    # Add gpu count if first iteration
+                    if iter == 0:
+                        count = dict(
+                            stat_name='gpu_count',
+                            gpu_type=gpu_type,
+                            gpu_index=gpu)
+
+                        count['stat_value'] = len(metric_ids)
+                        cache_stats.append(count)
+                iter += 1
+
+        return list(cache_stats)
+
+    def get_gpu_type_by_resource_id(self):
+        """Return dict of blazar hosts by hypervisor hostname."""
+        blazar_api = session_adapter('reservation')
+        hosts = blazar_api.get('os-hosts?detail=True').json()['hosts']
+
+        return {
+            h['hypervisor_hostname']: h['node_type']
+            for h in hosts
+            if h['gpu.gpu']}
+
+    def get_metrics_by_gpu_type(self):
+        """Return dict of metrics for resource type cuda from gnocchi."""
+        cuda_resouce_url = GNOCCHI_ENDPOINT + '/search/resource/cuda'
```

Comment:
I think you can probably get a gnocchi session by doing `session_adapter('metric')` or something. That would obviate the need for the `GNOCCHI_ENDPOINT` required env var.